lib/util/cache/repository/impl/s3.ts

```diff
@@ -0,0 +1,72 @@
+import { Readable } from 'stream';
+import {
+  GetObjectCommand,
+  GetObjectCommandInput,
+  PutObjectCommand,
+  PutObjectCommandInput,
+} from '@aws-sdk/client-s3';
+import { logger } from '../../../../logger';
+import { getS3Client, parseS3Url } from '../../../s3';
+import { streamToString } from '../../../streams';
+import type { RepoCacheRecord } from '../types';
+import { RepoCacheBase } from './base';
+
+export class RepoCacheS3 extends RepoCacheBase {
+  private readonly s3Client;
+  private readonly bucket;
+
+  constructor(repository: string, url: string) {
+    super(repository);
+    this.bucket = parseS3Url(url)?.Bucket;
+    this.s3Client = getS3Client();
+  }
+
+  async read(): Promise<string | undefined> {
+    const cacheFileName = this.getCacheFileName();
+    const s3Params: GetObjectCommandInput = {
+      Bucket: this.bucket,
+      Key: cacheFileName,
+    };
+    try {
+      const { Body: res } = await this.s3Client.send(
+        new GetObjectCommand(s3Params)
+      );
+      if (res instanceof Readable) {
+        const data: string = JSON.parse(await streamToString(res));
+        logger.debug('RepoCacheS3.read() - success'); // parse can throw
+        return data;
+      }
+      logger.warn(
+        `RepoCacheS3.read() - failure - expecting Readable return type got '${typeof res}' type instead`
+      );
+    } catch (err) {
+      // https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html
+      if (err?.name === 'NoSuchKey') {
+        logger.debug(`RepoCacheS3.read() - 'NoSuchKey'`);
+      } else {
+        logger.warn({ err }, 'RepoCacheS3.read() - failure');
+      }
+    }
+    return undefined;
+  }
+
+  async write(data: RepoCacheRecord): Promise<void> {
+    const cacheFileName = this.getCacheFileName();
+    const s3Params: PutObjectCommandInput = {
+      Bucket: this.bucket,
+      Key: cacheFileName,
+      Body: JSON.stringify(data),
+      ContentType: 'application/json',
+    };
+    try {
+      const res = await this.s3Client.send(new PutObjectCommand(s3Params));
+      logger.debug({ res }, 'RepoCacheS3.write() - success');
+    } catch (err) {
+      logger.warn({ err }, 'RepoCacheS3.write() - failure');
```

Comment:
`{ err }` will print the stacktrace as well, better to minimize the `err` and only log `err.code` and `err.message` since these are s3 documented errors